src/Equinor.ProCoSys.IPO.Command/InvitationCommands/UnCompletePunchOut/UnCompletePunchOutCommandHandler.cs

```diff
@@ -0,0 +1,84 @@
+using Equinor.ProCoSys.IPO.Domain.AggregateModels.InvitationAggregate;
+using Equinor.ProCoSys.IPO.Domain.AggregateModels.PersonAggregate;
+using Equinor.ProCoSys.IPO.ForeignApi.MainApi.McPkg;
+using Equinor.ProCoSys.IPO.ForeignApi.MainApi.Person;
+using MediatR;
+using ServiceResult;
+
+namespace Equinor.ProCoSys.IPO.Command.InvitationCommands.UnCompletePunchOut
+{
+    public class UnCompletePunchOutCommandHandler : IRequestHandler<UnCompletePunchOutCommand, Result<string>>
+    {
+        private readonly IPlantProvider _plantProvider;
+        private readonly IInvitationRepository _invitationRepository;
+        private readonly IUnitOfWork _unitOfWork;
+        private readonly ICurrentUserProvider _currentUserProvider;
+        private readonly IMcPkgApiService _mcPkgApiService;
+
+        public UnCompletePunchOutCommandHandler(
+            IPlantProvider plantProvider,
+            IInvitationRepository invitationRepository,
+            IUnitOfWork unitOfWork,
+            ICurrentUserProvider currentUserProvider, 
+            IMcPkgApiService mcPkgApiService)
+        {
+            _plantProvider = plantProvider;
+            _invitationRepository = invitationRepository;
+            _unitOfWork = unitOfWork;
+            _currentUserProvider = currentUserProvider;
+            _mcPkgApiService = mcPkgApiService;
+        }
+
+        public async Task<Result<string>> Handle(UnCompletePunchOutCommand request, CancellationToken cancellationToken)
+        {
+            var invitation = await _invitationRepository.GetByIdAsync(request.InvitationId);
+            var currentUserAzureOid = _currentUserProvider.GetCurrentUserOid();
+            var participant = invitation.Participants.SingleOrDefault(p => 
+                p.SortKey == 0 && 
+                p.Organization == Organization.Contractor && 
+                p.AzureOid == currentUserAzureOid);
+
+            if (participant == null || participant.FunctionalRoleCode != null)
+            {
+                var functionalRole = invitation.Participants
+                    .SingleOrDefault(p => p.SortKey == 0 &&
+                                          p.FunctionalRoleCode != null &&
+                                          p.Type == IpoParticipantType.FunctionalRole);
+
+                invitation.UnCompleteIpo(functionalRole, request.ParticipantRowVersion);
+            }
+            else
+            {
+                invitation.UnCompleteIpo(participant, request.ParticipantRowVersion);
+            }
+
+            invitation.SetRowVersion(request.InvitationRowVersion);
+            await ClearM01DatesAsync(invitation);
+            await _unitOfWork.SaveChangesAsync(cancellationToken);
+            return new SuccessResult<string>(invitation.RowVersion.ConvertToString());
+        }
+
+        private async Task ClearM01DatesAsync(Invitation invitation)
+        {
+            try
+            {
+                await _mcPkgApiService.ClearM01DatesAsync(
```

Comment:
We need to make changes to_mcPkgApiService.ClearM01DatesAsync  so that we clear externalReference in this case. I think this is used one other place where we do not clear external reference, hence this is the logic in the method. Add paramterer (optional), clearExternalReference, and clear it in this case